### PR TITLE
Implement hermite_renormalized_batched vjp

### PR DESCRIPTION
**Pull request by apchytr**
### **User description**
**Context:** Currently, the only hermite renorm custom gradient ported from Tensorflow is `hermite_renormalized_unbatched`. In this PR, `hermite_renormalized_batched` is ported.  

**Description of the Change:** Added `hermite_renormalized_batched_jax` `hermite_renormalized_batched_jax_fwd` and `hermite_renormalized_batched_jax_bwd` 

**Benefits:** Can now run optimizations on batched objects 

**Possible Drawbacks:** Some caveats apply with batched optimizations. Namely, the cost function still has to return a single scalar

___

### **PR Type**
Enhancement

___

### **Description**
- Implement custom JAX gradient for `hermite_renormalized_batched`

- Replace pure callback with custom VJP implementation

- Add forward and backward pass functions

- Enable gradient-based optimization for batched operations

___

### **Changes diagram**

```mermaid
flowchart LR
  A["hermite_renormalized_batched"] --> B["hermite_renormalized_batched_jax"]
  B --> C["hermite_renormalized_batched_jax_fwd"]
  B --> D["hermite_renormalized_batched_jax_bwd"]
  C --> E["Forward Pass"]
  D --> F["Gradient Computation"]
```

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>backend_jax.py</strong><dd><code>Replace batched hermite implementation with custom gradient</code></dd></summary>
<hr>

mrmustard/math/backend_jax.py

<li>Import <code>hermite_renormalized_batched_jax</code> function<br> <li> Replace pure callback implementation with custom VJP<br> <li> Simplify error message formatting

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/625/files#diff-277d6b9d9077473e0fca45c03a82a1503f283442e0b7503da968e5330ec3c242">+8/-18</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>hermite.py</strong><dd><code>Add batched hermite custom gradient implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/math/jax_vjps/hermite.py

<li>Add <code>hermite_renormalized_batched_jax</code> with custom VJP decorator<br> <li> Implement forward pass function <code>hermite_renormalized_batched_jax_fwd</code><br> <li> Implement backward pass function <code>hermite_renormalized_batched_jax_bwd</code><br> <li> Register VJP functions with <code>defvjp</code>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/625/files#diff-6aac5f7f8566a0219c57e5294e6c100fe128ac085710abde68a9bff8a0005cfd">+69/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: tests/test_training/test_opt_lab_jax.py [hunk 1]
```diff
@@ -209,51 +200,57 @@ def cost_fn(bsgate):
...
         """Test that the gradient of a squeezing gate is computed from the fock representation."""
-        squeezing = Sgate(0, r=1.0, r_trainable=True)
+        squeezing = Sgate(0, r=math.ones(batch_shape), r_trainable=True)
         og_r = math.asnumpy(squeezing.parameters.r.value)
         num = Number(0, 2)
         vac = Vacuum(0).dual
 
         def cost_fn(squeezing):
-            return -math.real((num >> squeezing >> vac) ** 2)
+            norm = 1 / squeezing.ansatz.batch_size if squeezing.ansatz.batch_shape else 1
```
[the remark this answers is not in the document]
Comment (apchytr) [reply tests/test_training/test_opt_lab_jax.py:212]: Yeah, `batch_shape` is the actual attribute we store and could be though of as the "source". `batch_size` on the other hand is just a convenience thing implemented as `math.prod(self.batch_shape) if self.batch_shape else 0`. This here is really only to make sure it's 1 and not 0 to avoid a division by zero error
